client/client_session.cpp

```diff
@@ -1,70 +1,151 @@
 #include "client_session.h"
 
-using tcp = boost::asio::ip::tcp;              // from <boost/asio/ip/tcp.hpp>
-namespace ssl = boost::asio::ssl;              // from <boost/asio/ssl.hpp>
-namespace websocket = boost::beast::websocket; // from <boost/beast/websocket.hpp>
+#include <boost/algorithm/string.hpp>
+#include <sstream>
+
+namespace ssl = boost::asio::ssl;       // from <boost/asio/ssl.hpp>
+namespace beast = boost::beast;         // from <boost/beast.hpp>
+namespace websocket = beast::websocket; // from <boost/beast/websocket.hpp>
 
 using namespace nibaclient;
 
-// , work_(ioc_)
 nibaclient::client_session::client_session(std::string const &host, std::string const &port,
-                                           boost::asio::io_context &ioc,
-                                           boost::asio::ssl::context &ssl_ctx) :
-    host_(host),
-    port_(port), ioc_(ioc), resolver_(ioc_), ws_(ioc_, ssl_ctx), timer_(ioc_) {
-    // everything can be sync because this client has no ui anyway
-    auto const results = resolver_.resolve(host_, port_);
-    boost::asio::connect(ws_.next_layer().next_layer(), results.begin(), results.end());
-    tcp::no_delay option(true);
-    ws_.next_layer().next_layer().set_option(option);
-    ws_.next_layer().handshake(ssl::stream_base::client);
-    ws_.handshake(host_, "/");
-    timer_.expires_after(std::chrono::seconds(5));
-    ping_timer({});
-    // do a data exchange
+                                           boost::asio::io_context &ioc, ssl::context &ssl_ctx) :
+    host_{host},
+    port_{port}, ioc_{ioc}, resolver_{ioc_}, ws_{ioc_, ssl_ctx}, ready_promise_{},
+    ready_future_{ready_promise_.get_future()} {}
+
+void nibaclient::client_session::start() {
+    // Note we only have 1 ioc_ thread, so no strand needed
+    boost::asio::spawn(ioc_, [this, self = shared_from_this()](boost::asio::yield_context yield) {
+        try {
```

Comment:
Is this try-catch block a one-time thing? That means ready_promise and ready_future are used just once during client connection setup?

Reply:
Yes, it's just for setting up the connection. I was making an attempt to eliminate exceptions and use error codes instead, but I got lazy and didn't care much for this part. Also, forcing an error via `set_exception(std::current_exception())` seemed a lot easier than figuring out how to do it properly.